### PR TITLE
WIP: Not-quite-working exec patch

### LINKED ISSUE #117: MismatchedTokenError caused by use of exec() function
The problem is that `ast` compiles both the exec statement and exec function into the same AST nodes, while patchedast.py assumes the `exec` statement.

``` python
>>> import ast
>>> ast.Exec
<class '_ast.Exec'>
>>> ast.Exec._fields
('body', 'globals', 'locals')
>>> ast.dump(ast.parse('exec "def f(): pass" in {}'))
"Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
>>> ast.dump(ast.parse('exec("def f(): pass", {})'))
"Module(body=[Exec(body=Str(s='def f(): pass'), globals=Dict(keys=[], values=[]), locals=None)])"
```

From `patchedast.py`, lines 335-342.

``` python

    def _Exec(self, node):
        children = []
        children.extend(['exec', node.body])
        if node.globals:
            children.extend(['in', node.globals])
        if node.locals:
            children.extend([',', node.locals])
        self._handle(node, children)
```

I'm willing to write a patch and pull request, but what's the right way to rewrite _Exec?

### PR DESCRIPTION
**Pull request by ceridwen**
I followed the pattern in _excepthandler as much as possible, but the tests are still failing, I suspect because the exec function has parentheses instead of white space, and I don't know how to handle the case where symbols are optional and can't find any analogous cases in the rest of the code.  Can you point me to what I need to add?

Fixes #117

### CODE REVIEW

**Comment by mcepl** [comment]
@aligrudi  Can I ask for help here, please?

File: rope/refactor/patchedast.py [hunk 1]
```diff
@@ -649,7 +693,8 @@ def consume(self, token):
                     break
                 else:
                     self._skip_comment()
-        except (ValueError, TypeError):
+        except (ValueError, TypeError) as e:
+            print e
```
Comment (emacsway) [inline rope/refactor/patchedast.py]: forgotten debug?

**Comment by mcepl** [comment]
I tried to rebase this patch on the top of the current master ([reformed patch](https://github.com/python-rope/rope/files/5918009/0001-Not-quite-working-exec-patch.patch.txt)), but I still have couple of tests failing:

```
======================================================================================= FAILURES =======================================================================================
__________________________________________________________ PatchedASTTest.test_call_func_and_both_varargs_and_kwargs_python35 __________________________________________________________

self = <ropetest.refactor.patchedasttest.PatchedASTTest testMethod=test_call_func_and_both_varargs_and_kwargs_python35>

    @testutils.only_for('3.5')
    def test_call_func_and_both_varargs_and_kwargs_python35(self):
        source = 'f(*args, **kwds)\n'
        ast_frag = patchedast.get_patched_ast(source, True)
        checker = _ResultChecker(self, ast_frag)
>       checker.check_children(
            'Call', ['Name', '', '(', '*', 'Starred', '', ',',
                     ' **', 'keyword', '', ')'])

ropetest/refactor/patchedasttest.py:469:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ropetest/refactor/patchedasttest.py:1195: in check_children
    self.test_case.assertEqual(len(children), len(result))
E   AssertionError: 11 != 7
_____________________________________________________________________________ PatchedASTTest.test_set_node _____________________________________________________________________________

self = <ropetest.refactor.patchedasttest.PatchedASTTest testMethod=test_set_node>

    def test_set_node(self):
        # make sure we are in a python version with set literals
        source = '{1, 2}\n'

        try:
            eval(source)
        except SyntaxError:
            return

        ast_frag = patchedast.get_patched_ast(source, True)
        checker = _ResultChecker(self, ast_frag)
>       checker.check_region('Set', 0, len(source) - 1)

ropetest/refactor/patchedasttest.py:796:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ropetest/refactor/patchedasttest.py:1165: in check_region
    self.test_case.assertEqual((start, end), node.region)
E   AssertionError: Tuples differ: (0, 6) != (0, 0)
E
E   First differing element 1:
E   6
E   0
E
E   - (0, 6)
E   ?     ^
E
E   + (0, 0)
E   ?     ^
_____________________________________________________________________ PatchedASTTest.test_starargs_after_keywords ______________________________________________________________________

self = <ropetest.refactor.patchedasttest.PatchedASTTest testMethod=test_starargs_after_keywords>

    @testutils.only_for('3.5')
    def test_starargs_after_keywords(self):
        source = 'foo(a=1, *args)\n'
        ast_frag = patchedast.get_patched_ast(source, True)
        checker = _ResultChecker(self, ast_frag)
>       checker.check_children(
            'Call', ['Name', '', '(', '', 'keyword', '', ',', ' *',
                     'Starred', '', ')'])

ropetest/refactor/patchedasttest.py:1150:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ropetest/refactor/patchedasttest.py:1195: in check_children
    self.test_case.assertEqual(len(children), len(result))
E   AssertionError: 11 != 7
_____________________________________________________________________ PatchedASTTest.test_starargs_before_keywords _____________________________________________________________________

self = <ropetest.refactor.patchedasttest.PatchedASTTest testMethod=test_starargs_before_keywords>

    @testutils.only_for('3.5')
    def test_starargs_before_keywords(self):
        source = 'foo(*args, a=1)\n'
        ast_frag = patchedast.get_patched_ast(source, True)
        checker = _ResultChecker(self, ast_frag)
>       checker.check_children(
            'Call', ['Name', '', '(', '*', 'Starred', '', ',', ' ',
                     'keyword', '', ')'])

ropetest/refactor/patchedasttest.py:1132:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ropetest/refactor/patchedasttest.py:1195: in check_children
    self.test_case.assertEqual(len(children), len(result))
E   AssertionError: 11 != 7
_______________________________________________________________________ PatchedASTTest.test_starargs_in_keywords _______________________________________________________________________

self = <ropetest.refactor.patchedasttest.PatchedASTTest testMethod=test_starargs_in_keywords>

    @testutils.only_for('3.5')
    def test_starargs_in_keywords(self):
        source = 'foo(a=1, *args, b=2)\n'
        ast_frag = patchedast.get_patched_ast(source, True)
        checker = _ResultChecker(self, ast_frag)
>       checker.check_children(
            'Call', ['Name', '', '(', '', 'keyword', '', ',', ' *',
                     'Starred', '', ',', ' ', 'keyword', '',')'])

ropetest/refactor/patchedasttest.py:1141:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
ropetest/refactor/patchedasttest.py:1195: in check_children
    self.test_case.assertEqual(len(children), len(result))
E   AssertionError: 15 != 7
```
@ceridwen , would you be able to finish this patch somehow? Thank you.

**Comment by lieryan** [comment]
Closing this PR in favor of a revised version of @ceridwen's PR on #372.

**Pull request closed without merging**